tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/RawTensorTest.java

```diff
@@ -0,0 +1,74 @@
+/*
+ *  Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *  =======================================================================
+ */
+package org.tensorflow;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import org.junit.jupiter.api.Test;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.types.TFloat32;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.types.TString;
+
+public class RawTensorTest {
+
+  @Test
+  public void rawToTypedTensor() {
+    RawTensor rawTensor = RawTensor.allocate(TFloat32.class, Shape.of(2, 2), -1);
+    TFloat32 floatTensor = rawTensor.asTypedTensor();
+    assertSame(floatTensor.asRawTensor(), rawTensor);
+    try {
+      TInt32 intTensor = rawTensor.asTypedTensor();
```

Comment:
Good point, I've never tried that before. You want me to update the doc or try to throw an `ClassCastException` in this case as well? 

Note that `asTypedTensor()` is an internal package-private method and where it is currently being used, it is implicitly enforced that the returned type matches the type of the tensor for all cases.

Reply:
Updating the doc seems good to me for both of the reasons you give.

Reply:
To be honest I'm not too sure how to document this. It looks to me that this behavior is some kind of "defect" in the generic specification, since the explicit `<TInt32>` parameterization of the method invocation seems to be overridden implicitly to the type inferred by the target (`Object` in this case). 

So basically, `<TInt32>` is ignored and that is probably the case for Java type inference in general when dealing with type parameters inferred by a target, is it worth documenting it here then?

Reply:
The type will be erased at runtime so the cast is erased to its bound (`TType`) which always succeeds. It might be better to have it explicitly return `TType` rather than have it promise something that can't be enforced by the type system. This is going to pollute the internals of our code with casts, but at least we'll remember to put them in rather than having it mysteriously blow up.

Reply:
Ok sounds fair, let's do that, anyway all these additional casts will happen internally only.

Reply:
On further reflection, I lean the same way as @Craigacp: have this method return `TType`.

Here's some analysis to support this choice: At runtime, the current behavior is "apply the mapper from `typeInfo` and return whatever `TType` it produces." Since the type parameter `T` is erased at runtime, the only way to guarantee a return of `T` would be to have the method take the tensor type class as a runtime parameter and explicitly verify that the mapper returns that type class. I might advocate this approach for a public method or even a widely used package-private method, but not for a rarely used package-private method.

But also, a review of the call points of this method raises another issue: whenever we treat the return value of `asTypedTensor` as simply a `Tensor` (rather than a `TType`), we "forget" at compile time the semantic upgrade that was presumably the point of the `asTypedTensor` call in the first place.

As an example, let's explore what happens underneath the following method:
~~~ Java
class EagerOperation extends AbstractOperation {
  // ...
  /**
   * Returns the tensor of the {@code outputIdx}th output of this operation.
   *
   * <p>This is only supported in an eager execution environment.
   *
   * @param outputIdx index of the output of this operation
   * @return output tensor
   */
  @Override
  Tensor tensor(int outputIndex) {
    Tensor tensor = outputTensors.get(outputIndex);
    if (tensor == null) {
      tensor = resolveTensor(outputIndex);
    }
    return tensor;
  }
  // ...
}
~~~
(I copied the Javadoc above the `@Override` for easy reference.)

The caller should presumably be agnostic as to whether this returns a `RawTensor` or a `TType`. But looking deeper into the call path, here's how the returned `Tensor` is  constructed:
~~~ Java
  private static Tensor resolveTensorHandle(TFE_TensorHandle handle, EagerSession session) {
    requireTensorHandle(handle);
    try (PointerScope scope = new PointerScope()) {
      TF_Status status = TF_Status.newStatus();
      TF_Tensor tensor = TFE_TensorHandleResolve(handle, status).withDeallocator();
      status.throwExceptionIfNotOK();
      return RawTensor.fromHandle(tensor, session).asTypedTensor();
    }
  }
~~~
What's the point of `asTypedTensor()` on the last substantive line, above? It causes this method to return a special kind of `Tensor` -- a `TType` -- but that fact is immediately forgotten by the type system and is not even asserted in the Javadoc. Presumably, we should either drop the call to `.asTypedTensor()` or change the whole call path to return `TType`.

Reply:
:-) But not for this PR!